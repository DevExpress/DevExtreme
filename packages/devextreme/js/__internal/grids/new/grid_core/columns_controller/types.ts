import type { ColumnBase } from '@js/common/grids';

export interface Column extends Pick<Required<ColumnBase>, 'alignment' | 'dataType' | 'visible' | 'caption' | 'visibleIndex'> {
  dataField?: string;

  name: string;

  calculateCellValue: (this: this, data: unknown) => unknown | Promise<unknown>;

  editorTemplate?: unknown;

  fieldTemplate?: unknown;
}

export type ColumnProperties = Partial<Column> | string;

export interface Cell {
  value: unknown;
  column: Column;
}

export interface DataRow {
  cells: Cell[];

  key: unknown;
}

export interface ColumnsControllerProperties {
  columns?: ColumnProperties[];
}
