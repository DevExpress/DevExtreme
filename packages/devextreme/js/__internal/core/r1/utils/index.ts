export {
  getTemplate,
} from './get_template';
export {
  shallowEquals,
} from './shallow_equals';
export {
  updatePropsImmutable,
} from './update_props_immutable';
