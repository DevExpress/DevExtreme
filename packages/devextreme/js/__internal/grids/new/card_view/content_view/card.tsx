import $ from '@js/core/renderer';
import type { DataRow } from '@ts/grids/new/grid_core/columns_controller/types';
import { PureComponent } from '@ts/grids/new/grid_core/core/pure_component';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { type ComponentType, type InfernoNode, render } from 'inferno';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import type { FieldProps } from './field';
import { Field } from './field';

export const CLASSES = {
  card: 'dx-cardview-card',
};

export interface CardProps {
  row: DataRow;
  isEditing?: boolean;

  onChange?: (columnName: string, value: unknown) => void;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  fieldTemplate?: any;
}

export class Card extends PureComponent<CardProps> {
  render(): InfernoNode {
    const FieldTemplate = this.props.fieldTemplate ?? Field;
    return (
      <div className={CLASSES.card} tabIndex={0}>
        {this.props.row.cells.map((cell, index) => (
          <FieldTemplate
            index={index}
            // eslint-disable-next-line max-len, @typescript-eslint/explicit-function-return-type
            defaultTemplate={{ render(model, _index, container) { render(<Field {...model} />, $(container).get(0)); } }}
            alignment={cell.column.alignment}
            title={cell.column.caption}
            value={cell.value}
            isEditing={this.props.isEditing}
            onChanged={(v): void => {
              if (!cell.column.name) {
                return;
              }
              this.props.onChange?.(cell.column.name, v);
            }}
          />
        ))}
      </div>
    );
  }
}
