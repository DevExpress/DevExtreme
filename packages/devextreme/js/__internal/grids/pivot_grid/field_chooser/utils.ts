import type { SortOrderType } from './const';
import { SORT_ORDER } from './const';

export const reverseSortOrder = (
  sortOrder: SortOrderType,
): SortOrderType => (sortOrder === SORT_ORDER.descending
  ? SORT_ORDER.ascending
  : SORT_ORDER.descending);
