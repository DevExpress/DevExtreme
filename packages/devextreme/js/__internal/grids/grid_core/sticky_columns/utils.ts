import { isDefined } from '@js/core/utils/type';

import type { ColumnsController } from '../columns_controller/m_columns_controller';
import { STICKY_BORDER_WIDTH, StickyPosition } from './const';

export const getColumnFixedPosition = (
  { fixedPosition }: { fixedPosition: StickyPosition | undefined },
): StickyPosition => fixedPosition ?? StickyPosition.Left;

const getStickyOffsetCore = function (
  columns,
  widths: number[],
  columnIndex: number,
  fixedPosition: StickyPosition,
  offsets?: Record<number, Record<string, number>>,
): number {
  const column = columns[columnIndex];
  const isChildColumn = isDefined(column.ownerBand);
  const targetColumnIsRight = fixedPosition === StickyPosition.Right;
  const targetColumnIsSticky = getColumnFixedPosition(column) === StickyPosition.Sticky;
  const processedColumns = targetColumnIsRight
    ? columns.slice(columnIndex + 1) : columns.slice(0, columnIndex).reverse();
  const processedWidths = targetColumnIsRight
    ? widths.slice(columnIndex + 1) : widths.slice(0, columnIndex).reverse();
  let offset = 0;
  let adjacentStickyColumnIndex = 0;
  let nonAdjacentStickyColumnCount = targetColumnIsSticky && processedColumns.length ? 1 : 0;

  processedColumns.forEach((col, colIndex: number) => {
    if (col.fixed && (!isDefined(offsets) || column.ownerBand === col.ownerBand)) {
      const columnIsSticky = getColumnFixedPosition(col) === StickyPosition.Sticky;
      const areNextOnlyFixedColumns = !processedColumns.slice(colIndex + 1)
        .some(({ fixed }: { fixed: boolean }) => !fixed);

      offset += processedWidths[colIndex];

      if (colIndex === 0 && areNextOnlyFixedColumns) {
        nonAdjacentStickyColumnCount = 0;
      } else if (targetColumnIsSticky && columnIsSticky && !areNextOnlyFixedColumns) {
        if (colIndex !== adjacentStickyColumnIndex) {
          nonAdjacentStickyColumnCount += 1;
          adjacentStickyColumnIndex = colIndex + 1;
        } else {
          adjacentStickyColumnIndex += 1;
        }
      }
    }
  });

  if (isChildColumn && isDefined(offsets)) {
    offset += offsets?.[column.ownerBand]?.[fixedPosition] ?? 0;

    return offset;
  }

  return offset - (nonAdjacentStickyColumnCount * STICKY_BORDER_WIDTH);
};

const isFirstOrLastColumn = function (
  that: ColumnsController,
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  column,
  rowIndex: number,
  onlyWithinBandColumn = false,
  isLast = false,
  fixedPosition?: StickyPosition,
): boolean {
  const rtlEnabled = that.option('rtlEnabled');
  const methodName = rtlEnabled !== isLast ? 'isLastColumn' : 'isFirstColumn';

  if (column.fixedPosition === StickyPosition.Sticky) {
    const parentColumn = that.getParentColumn(column) ?? column;

    if (that[methodName](parentColumn, 0)) {
      return false;
    }
  }

  return that[methodName](column, rowIndex, onlyWithinBandColumn, fixedPosition);
};

const prevColumnIsFixedCore = function (
  that: ColumnsController,
  column,
  visibleColumns,
): boolean {
  const visibleColumnIndex = that.getVisibleIndex(column.index, 0);
  const prevColumn = visibleColumns?.[visibleColumnIndex - 1];

  return prevColumn?.fixed as boolean
    && (!column.fixed
      || column.fixedPosition === StickyPosition.Sticky
      || column.fixedPosition !== prevColumn?.fixedPosition
    );
};

export const getStickyOffset = function (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  columns: any[],
  widths: number[],
  columnIndex: number,
  offsets?: Record<number, Record<string, number>>,
): Record<string, number> {
  const result: Record<string, number> = {};
  const column = columns[columnIndex];

  if (column) {
    const fixedPosition = getColumnFixedPosition(column);

    switch (fixedPosition) {
      case StickyPosition.Sticky: {
        const offsetLeft = getStickyOffsetCore(
          columns,
          widths,
          columnIndex,
          StickyPosition.Left,
          offsets,
        );

        const offsetRight = getStickyOffsetCore(
          columns,
          widths,
          columnIndex,
          StickyPosition.Right,
          offsets,
        );

        result.left = offsetLeft;
        result.right = offsetRight;
        break;
      }
      case StickyPosition.Right: {
        const offsetRight = getStickyOffsetCore(
          columns,
          widths,
          columnIndex,
          StickyPosition.Right,
          offsets,
        );

        result.right = offsetRight;
        break;
      }
      default: {
        const offsetLeft = getStickyOffsetCore(
          columns,
          widths,
          columnIndex,
          StickyPosition.Left,
          offsets,
        );

        result.left = offsetLeft;
      }
    }
  }

  return result;
};

export const prevColumnIsFixed = function (
  that: ColumnsController,
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  column,
  rowIndex: number,
): boolean {
  const parentColumn = that.getParentColumn(column);
  const visibleColumns = that.getVisibleColumns(0);

  if (parentColumn) {
    const isFirstColumn = that.isFirstColumn(column, rowIndex, true);

    if (isFirstColumn) {
      return prevColumnIsFixedCore(that, parentColumn, visibleColumns);
    }
  }

  return prevColumnIsFixedCore(that, column, visibleColumns);
};

export const normalizeOffset = function (offset: Record<string, number>): CSSStyleDeclaration {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const styleProps: any = {};

  if (isDefined(offset.left)) {
    styleProps.left = `${offset.left}px`;
  }

  if (isDefined(offset.right)) {
    styleProps.right = `${offset.right}px`;
  }

  return styleProps as CSSStyleDeclaration;
};

export const isFirstFixedColumn = function (
  that: ColumnsController,
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  column,
  rowIndex: number,
  onlyWithinBandColumn = false,
  fixedPosition?: StickyPosition,
): boolean {
  return isFirstOrLastColumn(that, column, rowIndex, onlyWithinBandColumn, false, fixedPosition);
};

export const isLastFixedColumn = function (
  that: ColumnsController,
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  column,
  rowIndex: number,
  onlyWithinBandColumn = false,
  fixedPosition?: StickyPosition,
): boolean {
  return isFirstOrLastColumn(that, column, rowIndex, onlyWithinBandColumn, true, fixedPosition);
};
