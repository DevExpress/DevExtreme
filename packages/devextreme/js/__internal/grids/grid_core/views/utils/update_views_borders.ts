/* eslint-disable @typescript-eslint/no-unsafe-return */
import type { dxElementWrapper } from '@js/core/renderer';
import { isDefined } from '@js/core/utils/type';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type View = any;

// TODO: Move to the grid_core/m_types views interface.
export interface ViewsWithBorder {
  columnHeadersView: View;
  rowsView: View;
  filterPanelView: View;
  footerView: View;
}

const CLASSES = {
  borderedTop: 'dx-bordered-top-view',
  borderedBottom: 'dx-bordered-bottom-view',
};

const getFirstVisibleViewElement = ({
  columnHeadersView,
  rowsView,
}: ViewsWithBorder): dxElementWrapper => {
  if (columnHeadersView?.isVisible()) {
    return columnHeadersView.element();
  }

  return rowsView.element();
};

const getLastVisibleViewElement = ({
  filterPanelView,
  footerView,
  rowsView,
}: ViewsWithBorder): dxElementWrapper => {
  if (filterPanelView?.isVisible()) {
    return filterPanelView.element();
  }

  if (footerView?.isVisible()) {
    return footerView.element();
  }

  return rowsView.element();
};

const getViewElementWithClass = (
  viewsWithBorder: ViewsWithBorder,
  className: string,
): dxElementWrapper | null => {
  const borderedView = Object.values(viewsWithBorder)
    .find((view) => view?.element()?.hasClass(className));

  return borderedView?.element() ?? null;
};

const shouldUpdateBorders = (
  viewName: string,
  viewsWithBorder: ViewsWithBorder,
): boolean => {
  if (!Object.keys(viewsWithBorder).includes(viewName)) {
    return false;
  }

  const { rowsView, ...otherViews } = viewsWithBorder;
  if (!isDefined(rowsView?.element?.())) {
    return false;
  }

  return Object.values(otherViews)
    .filter((view) => view?.isVisible?.())
    .every((view) => isDefined(view?.element()));
};

export const updateViewsBorders = (
  viewName: string,
  viewsWithBorder: ViewsWithBorder,
): void => {
  if (!shouldUpdateBorders(viewName, viewsWithBorder)) {
    return;
  }

  const $oldFirst = getViewElementWithClass(viewsWithBorder, CLASSES.borderedTop);
  const $oldLast = getViewElementWithClass(viewsWithBorder, CLASSES.borderedBottom);
  const $newFirst = getFirstVisibleViewElement(viewsWithBorder);
  const $newLast = getLastVisibleViewElement(viewsWithBorder);

  // @ts-expect-error The dxElementWrapper's "is" method is badly typed.
  if ($oldFirst && !$oldFirst.is($newFirst)) {
    $oldFirst.removeClass(CLASSES.borderedTop);
  }

  // @ts-expect-error The dxElementWrapper's "is" method is badly typed.
  if ($oldLast && !$oldLast.is($newLast)) {
    $oldLast.removeClass(CLASSES.borderedBottom);
  }

  if (!$newFirst.hasClass(CLASSES.borderedTop)) {
    $newFirst.addClass(CLASSES.borderedTop);
  }

  if (!$newLast.hasClass(CLASSES.borderedBottom)) {
    $newLast.addClass(CLASSES.borderedBottom);
  }
};
