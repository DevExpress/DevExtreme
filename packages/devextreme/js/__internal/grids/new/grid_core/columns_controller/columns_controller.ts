/* eslint-disable spellcheck/spell-checker */
import type { Gettable, Subscribable, Updatable } from '@ts/core/reactive/index';
import { computed, iif } from '@ts/core/reactive/index';

import { DataController } from '../data_controller/data_controller';
import { OptionsController } from '../options_controller/options_controller';
import type { Column, DataRow } from './types';
import { normalizeColumn } from './utils';

export class ColumnsController {
  private readonly columnsConfiguration = this.options.oneWay('columns');

  private readonly columnsFromDataSource = computed(
    (items: unknown[]) => {
      if (!items.length) {
        return [];
      }

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return Object.keys(items[0] as any);
    },
    [this.dataController.items],
  );

  public readonly columns = computed(
    (columnsConfiguration) => (columnsConfiguration ?? []).map(normalizeColumn),
    [
      iif(
        computed((columnsConfiguration) => !!columnsConfiguration, [this.columnsConfiguration]),
        this.columnsConfiguration,
        this.columnsFromDataSource,
      ),
    ],
  ) as unknown as Subscribable<Column[]> & Updatable<Column[]> & Gettable<Column[]>;

  public readonly visibleColumns = computed(
    (columns) => columns.filter((column) => column.visible),
    [this.columns],
  );

  public readonly nonVisibleColumns = computed(
    (columns) => columns.filter((column) => !column.visible),
    [this.columns],
  );

  public static dependencies = [OptionsController, DataController] as const;

  constructor(
    private readonly options: OptionsController,
    private readonly dataController: DataController,
  ) {}

  public createDataRow(data: unknown, columns: Column[]): DataRow {
    return {
      cells: columns.map((c) => ({
        column: c,
        value: c.calculateCellValue(data),
      })),
      key: this.dataController.getDataKey(data),
    };
  }
}
