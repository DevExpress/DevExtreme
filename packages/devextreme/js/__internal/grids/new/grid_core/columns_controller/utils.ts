/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { DataType } from '@js/common';
import { captionize } from '@js/core/utils/inflector';

import type { Column, ColumnProperties } from './types';

const defaultColumnProperties = {
  dataType: 'string',
  calculateCellValue(data): unknown {
    // @ts-expect-error
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return data[this.dataField!];
  },
  alignment: 'left',
  visible: true,
} satisfies Partial<Column>;

const defaultColumnPropertiesByDataType: Record<DataType, Exclude<ColumnProperties, string>> = {
  boolean: {
    alignment: 'center',
  },
  string: {

  },
  date: {

  },
  datetime: {

  },
  number: {
    alignment: 'right',
  },
  object: {

  },
};

export function normalizeColumn(column: ColumnProperties, index: number): Column {
  let col = column;

  if (typeof col === 'string') {
    col = { dataField: col };
  }

  const dataTypeDefault = defaultColumnPropertiesByDataType[
    col.dataType ?? defaultColumnProperties.dataType
  ];

  const name = col.name ?? col.dataField ?? `column${index}`;
  const caption = captionize(name);

  return {
    ...defaultColumnProperties,
    ...dataTypeDefault,
    name,
    caption,
    visibleIndex: index,
    ...col,
  };
}
