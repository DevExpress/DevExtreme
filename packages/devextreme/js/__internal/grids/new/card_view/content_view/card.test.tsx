import { expect, it } from '@jest/globals';
import { render } from 'inferno';

import { normalizeColumn } from '../../grid_core/columns_controller/utils';
import type { CardProps } from './card';
import { Card } from './card';

const testProps: CardProps = {
  row: {
    cells: [{
      value: 'qwe',
      column: normalizeColumn('asd', 1),
    }],
    key: 1,
  },
};

it('should be rendered', () => {
  const container = document.createElement('div');
  render(<Card row={testProps.row}/>, container);
  expect(container.innerHTML).toMatchSnapshot();
});
