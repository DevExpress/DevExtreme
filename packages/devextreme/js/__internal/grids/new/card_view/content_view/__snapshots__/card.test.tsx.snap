// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be rendered 1`] = `"<div class="dx-cardview-card" tabindex="0"><div class="dx-cardview-field" tabindex="0"><span class="dx-cardview-field-name">asd: </span><span class="dx-cardview-field-name" style="text-align: left;">qwe</span></div></div>"`;
